test/plugins/test-trace-generic-pool.js

```diff
@@ -0,0 +1,147 @@
+}
+
+var assert = require('assert');
+var common = require('./common');
+
+describe('generic-pool2', function() {
+  var ROOT_SPAN = 'root-span';
+  var CHILD_SPAN_1 = 'child-span-1';
+  var CHILD_SPAN_2 = 'child-span-2';
+
+  var api;
+  var genericPool;
+  before(function() {
+    api = require('../..').start({ samplingRate: 0, stackTraceLimit: 0 });
+    genericPool = require('./fixtures/generic-pool2');
+  });
+
+  after(function() {
+    common.stopAgent(api);
+  });
+
+  it('perserves context', function(done) {
+    var config = {
+      name: 'generic-pool2 test',
+      create: function(callback) {
+        callback(function() {
+          var childSpan = api.createChildSpan({ name: CHILD_SPAN_2 });
+          assert.ok(childSpan);
+          childSpan.endSpan();
+        });
+      },
+      destroy: function(fn) {
+      }
+    };
+
+    var pool = new genericPool.Pool(config);
+    api.runInRootSpan({ name: ROOT_SPAN }, function(span) {
+      pool.acquire(function(err, fn) {
+        assert.ifError(err);
+        var childSpan = api.createChildSpan({ name: CHILD_SPAN_1 });
+        assert.ok(childSpan);
+        fn();
+        childSpan.endSpan();
+        span.endSpan();
+
+        var spans = common.getTraces(api)[0].spans;
+        assert.ok(spans);
+        assert.strictEqual(spans.length, 3);
+        assert.strictEqual(spans[0].name, ROOT_SPAN);
+        assert.strictEqual(spans[1].name, CHILD_SPAN_1);
+        assert.strictEqual(spans[2].name, CHILD_SPAN_2);
+
+        done();
+      });
+    });
+  });
+});
+
+describe('generic-pool3', function() {
+  var agent;
+  var genericPool;
+  before(function() {
+    agent = require('../..').start({ samplingRate: 0, stackTraceLimit: 0 });
+    genericPool = require('./fixtures/generic-pool3');
+  });
+
+  after(function() {
+    common.stopAgent(agent);
+  });
+
+  it ('preserves context', function() {
+    var ROOT_SPAN = 'root-span';
+    var CHILD_SPAN_1 = 'child-span-1';
+    var CHILD_SPAN_2 = 'child-span-2';
+    var CHILD_SPAN_3 = 'child-span-3';
+
+    var factory = {
+      create: function() {
+        return new Promise(function(resolve, reject) {
+          resolve(function(input) {
+            assert.strictEqual(input, 'SomeInput');
+            var childSpan = agent.createChildSpan({ name: CHILD_SPAN_2 });
+            assert.ok(childSpan);
+            childSpan.endSpan();
+          });
+        });
+      },
+
+      destroy: function(fn) {
+        return new Promise(function(resolve) {
+          resolve();
+        });
+      }
+    };
+
+    var opts = {
+      max: 1,
+      min: 1
+    };
+
+    var pool = genericPool.createPool(factory, opts);
+
+    var promise;
+    agent.runInRootSpan({ name: ROOT_SPAN }, function(span) {
+      promise = pool.acquire().then(function(fn) {
+        var childSpan = agent.createChildSpan({ name: CHILD_SPAN_1 });
+        assert.ok(childSpan);
+        fn('SomeInput');
+        childSpan.endSpan();
+        span.endSpan();
```

Comment:
This won't affect test behavior but for readability could this root span be closed after both of its children have been closed?